blockchain/filedao/filedao.go

```diff
@@ -248,30 +248,106 @@ func (fd *fileDAO) PutBlock(ctx context.Context, blk *block.Block) error {
 		log.L().Error("Block already exists.", zap.Uint64("height", blk.Height()), log.Hex("hash", h[:]))
 		return ErrAlreadyExist
 	}
-	// TODO: check if need to split DB
+
+	// check if we need to split DB
+	if err := fd.prepNextDbFile(blk.Height()); err != nil {
+		return err
+	}
 	return fd.currFd.PutBlock(ctx, blk)
 }
 
+func (fd *fileDAO) prepNextDbFile(height uint64) error {
+	fd.lock.Lock()
+	defer fd.lock.Unlock()
+
+	tip, err := fd.currFd.Height()
+	if err != nil {
+		return err
+	}
+	if height != tip+1 {
+		return ErrInvalidTipHeight
+	}
+
+	split, err := fd.needToSplitDB(height)
+	if err != nil || !split {
+		return err
+	}
+	return fd.addNewV2File(height)
+}
+
+func (fd *fileDAO) needToSplitDB(height uint64) (bool, error) {
+	filename := fd.cfg.DbPath
+	if fd.topIndex > 0 {
+		filename = kthAuxFileName(fd.cfg.DbPath, fd.topIndex)
+	}
+
+	dat, err := os.Stat(filename)
+	if err != nil {
+		return false, err
+	}
+
+	sizeExceedLimit := uint64(dat.Size()) > fd.cfg.SplitDBSize()
+	if fd.v2Fd == nil {
+		// transition from legacy file to v2 file
+		sizeExceedLimit = sizeExceedLimit && fd.hu.IsPost(config.Hawaii, height)
+	}
+	return sizeExceedLimit, nil
+}
+
+func (fd *fileDAO) addNewV2File(height uint64) error {
+	// create a new v2 file
+	cfg := fd.cfg
+	cfg.DbPath = kthAuxFileName(cfg.DbPath, fd.topIndex+1)
+	v2, err := newFileDAOv2(height, cfg)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err == nil {
+			fd.currFd = v2
+			fd.topIndex++
+		}
+	}()
+
+	// add the new v2 file to existing v2 manager
+	ctx := context.Background()
+	if fd.v2Fd != nil {
+		if err = fd.v2Fd.AddFileDAO(v2, height); err != nil {
+			return err
+		}
+		err = v2.Start(ctx)
+		return err
+	}
+
+	// create v2 manager
+	fd.v2Fd, _ = newFileV2Manager([]*fileDAOv2{v2})
+	err = fd.v2Fd.Start(ctx)
+	return err
+}
+
 func (fd *fileDAO) DeleteTipBlock() error {
 	return fd.currFd.DeleteTipBlock()
 }
 
 // CreateFileDAO creates FileDAO according to master file
 func CreateFileDAO(legacy bool, gen genesis.Genesis, cfg config.DB) (FileDAO, error) {
 	fd := fileDAO{
-		hu: config.NewHeightUpgrade(&gen),
+		hu:  config.NewHeightUpgrade(&gen),
+		cfg: cfg,
 	}
 
 	fds := []*fileDAOv2{}
-	_, v2Files := checkAuxFiles(cfg.DbPath, FileV2)
+	v2Top, v2Files := checkAuxFiles(cfg.DbPath, FileV2)
 	if legacy {
 		legacyFd, err := newFileDAOLegacy(cfg)
 		if err != nil {
 			return nil, err
 		}
 		fd.legacyFd = legacyFd
+		fd.topIndex, _ = checkAuxFiles(cfg.DbPath, FileLegacyAuxiliary)
```

Comment:
for example, 2 legacy files  chain-00000001.db, chain-00000002.db
then fd.topIndex = 2